File: packages/sdk-ts/src/stream.ts
import type { ExecStreamEvent } from '@sandchest/contract'
import type { ExecResult } from './types.js'

/** Parse SSE events from a streaming Response. */
export async function* parseSSE<T>(response: Response): AsyncGenerator<T> {
  const reader = response.body!.getReader()
  const decoder = new TextDecoder()
  let buffer = ''

  try {
    while (true) {
      const { done, value } = await reader.read()
      if (done) break

      buffer += decoder.decode(value, { stream: true })
      const parts = buffer.split('\n\n')
      buffer = parts.pop()!

      for (const part of parts) {
        for (const line of part.split('\n')) {
          if (line.startsWith('data: ')) {
            const data = line.slice(6)
            if (data) {
              yield JSON.parse(data) as T
            }
          }
        }
      }
    }
  } finally {
    reader.releaseLock()
  }
}

/**
 * A streaming exec result. Implements AsyncIterable so you can
 * `for await` over events, or call `collect()` to wait for the full result.
 *
 * Single-use: the underlying stream is consumed on first iteration.
 */
export class ExecStream implements AsyncIterable<ExecStreamEvent> {
  readonly execId: string
  private readonly _generator: AsyncGenerator<ExecStreamEvent>

  /** @internal — Use `sandbox.exec(cmd, { stream: true })` instead. */
  constructor(execId: string, generator: AsyncGenerator<ExecStreamEvent>) {
    this.execId = execId
    this._generator = generator
  }

  [Symbol.asyncIterator](): AsyncIterator<ExecStreamEvent> {
    return this._generator
  }

  /** Consume the entire stream and return the aggregated ExecResult. */
  async collect(): Promise<ExecResult> {
    let stdout = ''
    let stderr = ''
    let exitCode = 0
    let durationMs = 0

    for await (const event of this) {
      switch (event.t) {
        case 'stdout':
          stdout += event.data
          break
        case 'stderr':
          stderr += event.data
          break
        case 'exit':
          exitCode = event.code
          durationMs = event.duration_ms
          break
      }
    }

    return { execId: this.execId, exitCode, stdout, stderr, durationMs }
  }
}

File: packages/sdk-py/sandchest/stream.py
"""SSE parsing and ExecStream for streaming exec results."""

from __future__ import annotations

import json
from collections.abc import Generator, Iterator
from typing import TYPE_CHECKING

from .types import ExecResult, ExecStreamEvent

if TYPE_CHECKING:
    import httpx


def parse_sse(response: httpx.Response) -> Generator[ExecStreamEvent, None, None]:
    """Parse SSE events from a streaming httpx Response."""
    buffer = ""
    for chunk in response.iter_text():
        buffer += chunk
        while "\n\n" in buffer:
            part, buffer = buffer.split("\n\n", 1)
            for line in part.split("\n"):
                if line.startswith("data: "):
                    data = line[6:]
                    if data:
                        yield json.loads(data)
    response.close()


class ExecStream:
    """A streaming exec result.

    Iterate over events with ``for event in stream``, or call
    ``stream.collect()`` to wait for the full result.

    Single-use: the underlying stream is consumed on first iteration.
    """

    def __init__(
        self,
        exec_id: str,
        generator: Generator[ExecStreamEvent, None, None],
    ) -> None:
        self.exec_id = exec_id
        self._generator = generator

    def __iter__(self) -> Iterator[ExecStreamEvent]:
        return self._generator  # type: ignore[return-value]

    def collect(self) -> ExecResult:
        """Consume the entire stream and return the aggregated ExecResult."""
        stdout = ""
        stderr = ""
        exit_code = 0
        duration_ms = 0

        for event in self:
            if event["t"] == "stdout":
                stdout += event["data"]
            elif event["t"] == "stderr":
                stderr += event["data"]
            elif event["t"] == "exit":
                exit_code = event["code"]
                duration_ms = event["duration_ms"]

        return ExecResult(
            exec_id=self.exec_id,
            exit_code=exit_code,
            stdout=stdout,
            stderr=stderr,
            duration_ms=duration_ms,
        )
